hinted handoff: introduce HTTP API for waiting for hints to be sent

piodul opened the pull request:
This PR reworks code introduced in another PR: "hints: delay repair until hints are replayed" (#8452) and turns it into an HTTP API, completely separate from repair.

A concept of a hint waiting point is introduced. It can be used to wait for all currently accumulated hints on a set of nodes to be replayed towards another set of nodes.

A new HTTP endpoint for managing hint sync points is introduced: `/storage_proxy/hints_waiting_point`. It supports the following methods:

- `POST` - creates a new hint waiting point which waits for current hints to be sent from `source_hosts` to `target_hosts`.
- `GET` - gets the status of the waiting point or waits for a specified amount of time until the waiting point is reached.
- `DELETE` - deletes given hint waiting point and aborts the wait operation. Waiting points are not deleted automatically, they should be deleted by the user.

The PR starts by reverting a bunch of commits from the old implementation (11 commits at the end, with 9d68824327e607194207b6743565229a2862c252 skipped), leaving only the implementation of syncing with hints in `hints_manager` from the old PR. Then, commits with the implementation of new RPC, coordinating a hint waiting operation and the HTTP API follows.

Refs #8102
Fixes #8727

Tests: unit(dev)

piodul commented:
I see that there are some conflicts, will rebase.

I also wrote a dtest, but I'll send a separate PR which adds necessary error injection.

piodul commented:
@avikivity @asias @vladzcloudius could you take a look?

psarna commented:
> This PR reworks code introduced in another PR: "hints: delay repair until hints are replayed" (#8452) and turns it into an HTTP API, completely separate from repair.
> 
> A concept of a hint waiting point is introduced. It can be used to wait for all currently accumulated hints on a set of nodes to be replayed towards another set of nodes.
> 
> A new HTTP endpoint for managing hint sync points is introduced: `/storage_proxy/hints_waiting_point`. It supports the following methods:
> 
> * `POST` - creates a new hint waiting point which waits for current hints to be sent from `source_hosts` to `target_hosts`.
> * `GET` - gets the status of the waiting point or waits for a specified amount of time until the waiting point is reached.
> * `DELETE` - deletes given hint waiting point and aborts the wait operation. Waiting points are not deleted automatically, they should be deleted by the user.

Design comment: we absolutely need DoS protection here, otherwise it would be trivial to cause OOM with a bash script based on curl:
 1. There should be a policy of dropping old waiting points, i.e. they should have a TTL
 2. There should be a hard limit on waiting points, past which it's no longer possible to create more of them

> 
> The PR starts by reverting a bunch of commits from the old implementation (11 commits at the end, with [9d68824](https://github.com/scylladb/scylla/commit/9d68824327e607194207b6743565229a2862c252) skipped), leaving only the implementation of syncing with hints in `hints_manager` from the old PR. Then, commits with the implementation of new RPC, coordinating a hint waiting operation and the HTTP API follows.
> 
> Refs #8102
> Fixes #8727
> 
> Tests: unit(dev)

asias commented:
> @avikivity @asias @vladzcloudius could you take a look?

Sure. In my queue.

piodul commented:
> > This PR reworks code introduced in another PR: "hints: delay repair until hints are replayed" (#8452) and turns it into an HTTP API, completely separate from repair.
> > A concept of a hint waiting point is introduced. It can be used to wait for all currently accumulated hints on a set of nodes to be replayed towards another set of nodes.
> > A new HTTP endpoint for managing hint sync points is introduced: `/storage_proxy/hints_waiting_point`. It supports the following methods:
> > 
> > * `POST` - creates a new hint waiting point which waits for current hints to be sent from `source_hosts` to `target_hosts`.
> > * `GET` - gets the status of the waiting point or waits for a specified amount of time until the waiting point is reached.
> > * `DELETE` - deletes given hint waiting point and aborts the wait operation. Waiting points are not deleted automatically, they should be deleted by the user.
> 
> Design comment: we absolutely need DoS protection here, otherwise it would be trivial to cause OOM with a bash script based on curl:
> 
>     1. There should be a policy of dropping old waiting points, i.e. they should have a TTL
> 
>     2. There should be a hard limit on waiting points, past which it's no longer possible to create more of them
> 

@psarna Thanks for pointing the problem out, I should have considered it :)

Hmm... There are two kinds of resources here: waiting points on the coordinator (the node which handles the HTTP request), and futures which wait for hints to be replayed on each node participating in the operation (spawned by the HINT_SYNC verb). While I can implement what you propose for hint waiting points, the number of futures mentioned in the previous sentence should probably be limited in a different way.

Some context: previously, waiter <-> nodes replaying hints (let's call them followers) communication was handled through HINT_SYNC_POINT_CREATE and HINT_SYNC_POINT_CHECK verbs. The coordinator used the first verb to create a follower-local sync point and then checked the status by polling the second one every 1s. In this PR, I switched to using single synchronous HINT_SYNC verb because I thought it would result in simpler code.

Considering the possibility of memory overload, the old design should probably be revisited (or the HINT_SYNC verb could handle both operations of the old verbs).

I think we could resolve the problem of overload on the followers like this: follower-local sync points would have a relatively short TTL (e.g. 10 seconds) after which they expire and are deleted. Every time HINT_SYNC_POINT_CHECK is received, which should happen every 1 second, the TTL is reset.

If somebody is particularly malicious, they can theoretically still trigger OOM if they create and delete waiting points in a tight loop. To prevent this, we should probably also limit the number of the follower-local waiting points and fail further HINT_SYNC_POINT_CREATE requests until the number of the local waiting points drops.

> 
> > The PR starts by reverting a bunch of commits from the old implementation (11 commits at the end, with [9d68824](https://github.com/scylladb/scylla/commit/9d68824327e607194207b6743565229a2862c252) skipped), leaving only the implementation of syncing with hints in `hints_manager` from the old PR. Then, commits with the implementation of new RPC, coordinating a hint waiting operation and the HTTP API follows.
> > Refs #8102
> > Fixes #8727
> > Tests: unit(dev)

vladzcloudius commented:
@piodul I'll wait with a review till @psarna comments are addressed.

piodul commented:
@avikivity @asias In order to implement @psarna's suggestions I'll have to do some significant changes, so please wait with review until I send a v2.

asias commented:
> @avikivity @asias In order to implement @psarna's suggestions I'll have to do some significant changes, so please wait with review until I send a v2.

okay.

slivne commented:
I guess I am out of the loop - but a new distributed algorithm added at this point is not a great choice - is there a description of what we are trying to solve and how (I assune there is as everyone is overall ok with this direction or ask the same question that I have) - if there is can you point me to it (it may be a discussion on email thread / github issue.

Please note that repair probably had the same issue - a node needs to be repaired (or hints need to be caught up) - the management algorithm of knowing that the node is repaired is handled by the repairing node - although I did not see the details I expect the same can be done with hints - e.g. the node that has to catch up (how does he know that ?) - will manage the process and once he is done will update all I am done. This has the benefit of having a localized algorithm on the node that needs to catch up - its a pig change of pulling hints instead of pushing them - but it seems we are doing a big change already :)

piodul commented:
> I guess I am out of the loop - but a new distributed algorithm added at this point is not a great choice - is there a description of what we are trying to solve and how (I assune there is as everyone is overall ok with this direction or ask the same question that I have) - if there is can you point me to it (it may be a discussion on email thread / github issue.

@slivne 

For more context, you can take a look at the [issue](https://github.com/scylladb/scylla/issues/8727) relevant to this PR. It contains a short motivation of the idea behind this PR. The idea was discussed under the PR which introduced the "wait for hints before repair" feature (proposed here [here](https://github.com/scylladb/scylla/pull/8452#issuecomment-839458521)).

As for the description of the algorithm, it looks like I didn't describe it anywhere - I should probably add it to the developer documentation. Meanwhile, here is a short description.

Two RPC verbs are used:

- HINT_SYNC_POINT_CREATE - this verb contains a list of node IPs. When this verb is received, a "sync point" is created. A sync point tracks progress of hints on the local node towards nodes from the list. When all tracked hints are sent out, the point is considered "resolved".
- HINT_SYNC_POINT_CHECK - checks if given sync point has resolved or not.

The algorithm is simple - the coordinator (node handling the HTTP request) does the following:

1. Send HINTS_SYNC_POINT_CREATE to all nodes which we want to replay hints,
2. Periodically send HINT_SYNC_POINT_CHECK to nodes, and stop when all nodes report that their sync point has resolved.
3. At this point, hints are considered to be replayed.

This is the same algorithm and RPC as used in the "wait for hints before repair" feature, which was merged - it's just that users will be able to use it through HTTP API. Although I mention in the PR description that I'm making some changes to the RPC, I'll send a v2 which will not change the RPC and will work as described here.

> 
> Please note that repair probably had the same issue - a node needs to be repaired (or hints need to be caught up) - the management algorithm of knowing that the node is repaired is handled by the repairing node - although I did not see the details I expect the same can be done with hints - e.g. the node that has to catch up (how does he know that ?) - will manage the process and once he is done will update all I am done. This has the benefit of having a localized algorithm on the node that needs to catch up - its a pig change of pulling hints instead of pushing them - but it seems we are doing a big change already :)

The coordinator of the operation already does the tracking - it periodically asks nodes which are supposed to replay the hints if they are done. However, nodes which replay hints also have to do some tracking in order to reply to the coordinator (the "sync point" idea).

Maybe if we go with Avi's ([implied](https://github.com/scylladb/scylla/pull/8452#discussion_r617507373)) suggestion to use replay_position or something similar for ordering hints segments, we could get rid of the "sync point" thing on the nodes replaying hints, and those nodes won't have to keep state. However, I'd rather use a stateful solution here because it's what was implemented in the previous PR and should be easier to adapt, and getting rid of the follower state could be done in a followup.

As for converting hinted handoff from being push-based to pull-based, it sounds like a lot of effort, and the benefits are not clear to me.

Small side note: this algorithm is more generic than "replay all hints towards one node" case - in a single operation, you can provide sets of nodes A and B and ask nodes from set A to replay to nodes from B.

piodul commented:
v2: a major rework

- The old distributed algorithm is used (same RPC)
- Most of the hint sync point logic is moved into a new, separate service - sync_point_service
- Hint sync points expire after a 5-minute timeout (but can be deleted by the user earlier)

Sorry for the big change, but considering the need for additional logic related to limiting hint sync point count and their expiry, I thought it would be well to move the coordinator and follower logic into a separate service. Comments in the sync_point_service.hh file should be helpful in understanding how the new service works.

piodul commented:
@avikivity @asias @vladzcloudius I addressed @psarna 's comments and added some overload protection, I think it is OK to review now

psarna commented on db/hints/sync_point_service.cc:
```diff
@@ -0,0 +1,431 @@
+        }
+
+        // Waited on by the shard-local gate
+        (void)sharded_abort_source.invoke_on(0, [&] (auto& local0) {
```
This comment applies to this whole series: implicit capturing by reference is dangerous, especially for continuations - let's spend a while and explicitly state what's actually captured. In most places it's probably enough to just capture `this`, or even nothing

piodul replied:
Done, I looked for all `[&]`, `[&,` and `&]`'s and changed them to explicit captures.

piodul commented:
v2.1: converted implicit captures to explicit captures and added a little bit more of logging

piodul commented:
@avikivity @asias @vladzcloudius @psarna ping

piodul commented:
@avikivity @asias @vladzcloudius @psarna ping^2

There is a minor conflict related to changes in imports, I'll rebase and fix that after initial round of review.

piodul commented:
@avikivity @asias @vladzcloudius @psarna ping^3

avikivity commented on repair/repair.cc:
```diff
@@ -29,7 +29,6 @@
 #include "streaming/stream_reason.hh"
 #include "gms/inet_address.hh"
```
Every revert has to explain why it is reverting.

piodul replied:
Done, added this information to every revert commit.

avikivity commented on db/config.cc:
```diff
@@ -715,7 +715,6 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , override_decommission(this, "override_decommission", value_status::Used, false, "Set true to force a decommissioned node to join the cluster")
     , enable_repair_based_node_ops(this, "enable_repair_based_node_ops", liveness::LiveUpdate, value_status::Used, false, "Set true to use enable repair based node operations instead of streaming based")
-    , wait_for_hint_replay_before_repair(this, "wait_for_hint_replay_before_repair", liveness::LiveUpdate, value_status::Used, true, "If set to true, the cluster will first wait until the cluster sends its hints towards the nodes participating in repair before proceeding with the repair itself. This reduces the amount of data needed to be transferred during repair.")
     , ring_delay_ms(this, "ring_delay_ms", value_status::Used, 30 * 1000, "Time a node waits to hear from other nodes before joining the ring in milliseconds. Same as -Dcassandra.ring_delay_ms in cassandra.")
```
This changes a user-facing API. So the changelog has to explain why it is safe.

piodul replied:
Information added to the commit description (the option becomes useless and it's not a part of any release)

avikivity commented on db/hints/manager.cc:
```diff
@@ -40,7 +40,6 @@
 #include "database.hh"
 #include "service_permit.hh"
```
If you revert an entire merge, we can revert the merge commit. But we still need to explain it.

piodul replied:
Not everything is reverted. Unless I missed something, those are parts which are kept:

- RPC verbs are kept because they are used in this PR,
- The machinery which allows the hints manager to wait until hints are replayed - although it is modified a bit by commit https://github.com/scylladb/scylla/pull/8728/commits/2059077d5cf94765fa767e8b1e0520841b4d2bdb,
- Some small parts in repair code is now used by new code unrelated to this feature (e.g. get_hosts_participating_in_repair function, abort sources).

avikivity replied:
Ok.

avikivity commented on db/hints/sync_point_service.cc:
```diff
@@ -0,0 +1,442 @@
+                return with_scheduling_group(sps._task_scheduling_group, [&sps, &as, sync_point_id, target_endpoints = std::move(target_endpoints)] () mutable {
+                    return sps.wait_for_local_hints(as, sync_point_id, target_endpoints);
+                });
+            });
```
I don't understand this coordinator/follower thing. Does creating a sync point on the coordinator create sync point on all nodes? What if topology changes during this? What if a node restarts?

piodul replied:
The coordinator creates follower sync points only on the nodes from the `source_hosts` list. If one of the nodes from this list dies/is removed, the whole operation will fail.

avikivity commented on db/hints/sync_point_service.hh:
```diff
@@ -0,0 +1,302 @@
+    /// Coordinator tasks are scheduled for expiry after the internal function stops
+    /// (see `completed_coordinator_task_ttl`).
+    ///
+    /// Tasks should be created on task 0 only.
```
shard 0?

Maybe id % smp::count to distribute things.

piodul replied:
We are enforcing a limit on the number of tasks on a node, so this needs to be tracked on a single shard.

avikivity commented on db/hints/sync_point_service.cc:
```diff
@@ -0,0 +1,442 @@
+        auto& task = it->second;
+        task.run([&sps, &task, id, &source_nodes, &target_nodes] (abort_source& as) mutable {
+            return with_scheduling_group(sps._task_scheduling_group, [&sps, &task, &as, id, source_nodes = std::move(source_nodes), target_nodes = std::move(target_nodes)] () mutable {
+                return sps.coordinate_hint_waiting(as, id, std::move(source_nodes), std::move(target_nodes)).finally([&task] {
```
Why these scheduling group games? Shouldn't everything run under the maintenance group?

piodul replied:
`_task_scheduling_group` is the maintenance group. The `create_sync_point` function will be called by HTTP handler and I don't know which scheduling group it uses, so I put `with_scheduling_group` there just to be sure.

For the `HINT_SYNC_POINT_CREATE` verb handler, I agree - `with_scheduling_group` is redundant there because the handler will run in the scheduling. I put it there to be consistent with `create_sync_point`.

avikivity replied:
The REST API runs under the maintenance group.

avikivity replied:
```c++
            with_scheduling_group(maintenance_scheduling_group, [&] {
                return ctx.http_server.listen(socket_address{ip, api_port});
            }).get();
```

avikivity commented on db/hints/sync_point_service.cc:
```diff
@@ -0,0 +1,442 @@
+            if (!combined_as.abort_requested()) {
+                combined_as.request_abort();
+            }
+            throw;
```
What if a node is restarting?

piodul replied:
Then the whole waiting operation will fail and report a "FAILED" status.

avikivity replied:
And already-created sync points will be garbage collected by timeouts?

piodul replied:
Follower sync points are GC'd 10 seconds after the last time they were queried.

Coordinator queries all follower sync points for this operation every second. If one node becomes unreachable, the coordinator will report an error and stop querying followers. After 10 seconds follower sync points should be GC'd.

avikivity commented on db/hints/sync_point_service.cc:
```diff
@@ -0,0 +1,442 @@
+
+        return parallel_for_each(sps._managers, [&sharded_abort_source, &target_nodes, id] (manager& mgr) {
+            service_logger.debug("Task ({} | follower | shard {}): going to wait on shard for the following nodes: {}", this_shard_id(), id, target_nodes);
+            return mgr.wait_until_hints_are_replayed(sharded_abort_source.local().as, target_nodes);
```
"hints are replayed" is something that might never be true, if more hints are generated. Maybe we need to wait until all hints that were written before we started waiting are replayed.

piodul replied:
> Maybe we need to wait until all hints that were written before we started waiting are replayed.

This is the current semantic. I can change the name to something more precise (e.g. `wait_until_current_hints_are_replayed`?)

piodul replied:
Changed the name to `wait_until_current_hints_are_replayed`

avikivity commented:
I'd like to see a high level description of what we will do with this (in the cover letter).

piodul commented:
v3:
- Reworked the cover letter
- Added information to every revert commit why each of them is needed
- Renamed `wait_until_hints_are_replayed` to `wait_until_current_hints_are_replayed`

piodul commented:
It has dawned on me that this feature could probably be implemented without keeping any distributed state. Instead of using a UUID as a hint waiting point identifier, we could encode something like a replay position of all hint queues that we want to wait for.

The operation would be done through two HTTP endpoints:

1. Create a hint sync point - nodes from the `source_hosts` list are asked to go through their hint queues towards nodes in the `target_hosts` set and return something like a replay position indicating the end of the queues. Those replay positions are then compiled into a list, encoded and returned back to the user.
2. Check the hint sync point - nodes from the `source_hosts` list are asked if they reached relevant replay positions. If all of them did, success is returned, if not, we tell the client to wait. This could probably be made into a long poll.

Benefits include:

- No need to keep any state,
- No need for complicated overload protection.
- User is free to run operations 1. and 2. on different nodes each time,
- The operation is resilient to temporary failures such as node restarts (user may still get a failure but can continue waiting after the problem is resolved),

Drawbacks:

- The description of a hint sync point can be quite large if there are many nodes and shards involved which may result in lots of network traffic and long decoding of encoded positions `((number of nodes)^2 * shard count)`. This can be alleviated by doing long polls instead of short ones by the client.

I think it can be done, provided that hints manager can make use of the replay_position (I think it can).

@avikivity @vladzcloudius @asias @psarna @haaawk  opinions? I think this approach would be simpler and the lack of distributed state is very alluring, so I would like to try it unless you have any objections.

avikivity commented:
I would be very relieved to see a non-distributed implementation here. The current implementation looks like it will work, but the cost/benefit ratio is bad due to the high complexity of the implementation, in return for a very minor feature. So I'd welcome a simpler implementation.

In the new implementation, we can consider making the API per-node. Then the user has the responsibility of calling it on all nodes. This makes each sync point smaller, but puts a larger burden on the caller.

psarna commented:
@piodul sounds very good, especially with the per-node idea above applied

piodul commented:
Opened another PR with the stateless approach: https://github.com/scylladb/scylla/pull/8982. I'll close this one.
